cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -155,13 +158,34 @@ runQueryTip (AnyConsensusModeParams cModeParams) network mOutFile = do
   anyEra <- determineEra cModeParams localNodeConnInfo
   mEpoch <- mEpochQuery anyEra consensusMode  localNodeConnInfo
   tip <- liftIO $ getLocalChainTip localNodeConnInfo
+
+  let timeInterpreter = mkTimeInterpreter @IO
```

Comment:
I think the anticipated path is to do something similar to what `mEpochQuery` already does, but submitting the `GetInterpreter` query instead of `QueryEpoch`.

https://github.com/input-output-hk/ouroboros-network/blob/dea09273c87578a2047445e4434ba30df6570ac2/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger/Query.hs#L312

But now that I'm looking more closely, I see there's a whole other query data type in the `cardano-api` that I wasn't expecting. So I'm less sure about what path to actually suggest. I'll ping Duncan on our Slack thread.

Summary: you almost certainly want to get the `Interpreter` by querying the local node, and the Consensus-defined `GetInterpreter` is ultimately the only query I know for doing that.

Reply:
In attempting to add `GetInterpreter` here: https://github.com/input-output-hk/cardano-node/blob/3c556c5b09811c159dc7a92021f4d1c6aac3a104/cardano-api/src/Cardano/Api/Query.hs#L117

I would need to update `toConsensusQueryShelleyBased` https://github.com/input-output-hk/cardano-node/blob/3c556c5b09811c159dc7a92021f4d1c6aac3a104/cardano-api/src/Cardano/Api/Query.hs#L289

Does that mean I need extra constructors here?

https://github.com/input-output-hk/ouroboros-network/blob/ec086166c5492b1a407d8ea20b4e8c18589046cf/ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs#L82

Reply:
I can see [this definition](https://github.com/input-output-hk/ouroboros-network/blob/dea09273c87578a2047445e4434ba30df6570ac2/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger/Query.hs#L312):

```
data QueryHardFork xs result where
  GetInterpreter :: QueryHardFork xs (History.Interpreter xs)
```



data QueryHardFork xs result where
312 | GetInterpreter :: QueryHardFork xs (History.Interpreter xs)
…




data QueryHardFork xs result where
312 | GetInterpreter :: QueryHardFork xs (History.Interpreter xs)
…




data QueryHardFork xs result where
312 | GetInterpreter :: QueryHardFork xs (History.Interpreter xs)
…




data QueryHardFork xs result where
312 | GetInterpreter :: QueryHardFork xs (History.Interpreter xs)
…




data QueryHardFork xs result where
312 | GetInterpreter :: QueryHardFork xs (History.Interpreter xs)
…




data QueryHardFork xs result where
312 | GetInterpreter :: QueryHardFork xs (History.Interpreter xs)
…

Reply:
[here](https://github.com/input-output-hk/cardano-node/blob/3b5cc66ef7607fc4cd85d3276eca7b47b913f7a1/cardano-api/src/Cardano/Api/Query.hs#L274)

```
toConsensusQuery (QueryCurrentEra CardanoModeIsMultiEra) =
    Some (Consensus.QueryHardFork Consensus.GetCurrentEra)
```

[here](https://github.com/input-output-hk/cardano-node/blob/3b5cc66ef7607fc4cd85d3276eca7b47b913f7a1/cardano-api/src/Cardano/Api/IPC.hs#L448)

```
convLocalStateQueryClient mode =
    Net.Query.mapLocalStateQueryClient
      (toConsensusPointInMode mode)
      toConsensusQuery
      fromConsensusQueryResult
```

[here](https://github.com/input-output-hk/cardano-node/blob/3b5cc66ef7607fc4cd85d3276eca7b47b913f7a1/cardano-api/src/Cardano/Api/IPC.hs#L394)

```
convLocalNodeClientProtocols
    mode
    LocalNodeClientProtocols {
      localChainSyncClient,
      localTxSubmissionClient,
      localStateQueryClient
    } =
    LocalNodeClientProtocolsForBlock {
      localChainSyncClientForBlock    = case localChainSyncClient of
        NoLocalChainSyncClient -> NoLocalChainSyncClient
        LocalChainSyncClientPipelined clientPipelined -> LocalChainSyncClientPipelined $ convLocalChainSyncClientPipelined mode clientPipelined
        LocalChainSyncClient client -> LocalChainSyncClient $ convLocalChainSyncClient mode client,

      localTxSubmissionClientForBlock = convLocalTxSubmissionClient mode <$>
                                          localTxSubmissionClient,

      localStateQueryClientForBlock   = convLocalStateQueryClient mode <$>
                                          localStateQueryClient
    }
```

[here](https://github.com/input-output-hk/cardano-node/blob/3b5cc66ef7607fc4cd85d3276eca7b47b913f7a1/cardano-api/src/Cardano/Api/IPC.hs#L344)

```
mkLocalNodeClientParams :: forall mode block.
                           ConsensusBlockForMode mode ~ block
                        => ConsensusModeParams mode
                        -> LocalNodeClientProtocolsInMode mode
                        -> LocalNodeClientParams
mkLocalNodeClientParams modeparams clients =
    -- For each of the possible consensus modes we pick the concrete block type
    -- (by picking the appropriate 'Consensus.ProtocolClient' value).
    --
    -- Though it is not immediately visible, this point where we use
    -- 'LocalNodeClientParams' is also where we pick up the necessary class
    -- instances. This works because in each case we have a monomorphic block
    -- type and the instances are all in scope. This is why the use of
    -- LocalNodeClientParams is repeated within each branch of the case:
    -- because it is only within each branch that the GADT match makes the
    -- block type monomorphic.
    --
    case modeparams of
      ByronModeParams epochSlots ->
        LocalNodeClientParams
          (Consensus.ProtocolClientByron epochSlots)
          (convLocalNodeClientProtocols ByronMode clients)

      ShelleyModeParams ->
        LocalNodeClientParams
          Consensus.ProtocolClientShelley
          (convLocalNodeClientProtocols ShelleyMode clients)

      CardanoModeParams epochSlots ->
        LocalNodeClientParams
          (Consensus.ProtocolClientCardano epochSlots)
          (convLocalNodeClientProtocols CardanoMode clients)
```

[here](https://github.com/input-output-hk/cardano-node/blob/3b5cc66ef7607fc4cd85d3276eca7b47b913f7a1/cardano-api/src/Cardano/Api/IPC.hs#L179)

```
connectToLocalNode LocalNodeConnectInfo {
                     localNodeSocketPath,
                     localNodeNetworkId,
                     localConsensusModeParams
                   } clients =
    Net.withIOManager $ \iomgr ->
      Net.connectTo
        (Net.localSnocket iomgr localNodeSocketPath)
        Net.NetworkConnectTracers {
          Net.nctMuxTracer       = nullTracer,
          Net.nctHandshakeTracer = nullTracer
        }
        versionedProtocls
        localNodeSocketPath
  where
    versionedProtocls =
      -- First convert from the mode-parametrised view of things to the
      -- block-parametrised view and then do the final setup for the versioned
      -- bundles of mini-protocols.
      case mkLocalNodeClientParams localConsensusModeParams clients of
        LocalNodeClientParams ptcl clients' ->
          mkVersionedProtocols localNodeNetworkId ptcl clients'
```

[here](https://github.com/input-output-hk/cardano-node/blob/3b5cc66ef7607fc4cd85d3276eca7b47b913f7a1/cardano-api/src/Cardano/Api/IPC.hs#L463)

```
queryNodeLocalState :: forall mode result.
                       LocalNodeConnectInfo mode
                    -> Maybe ChainPoint
                    -> QueryInMode mode result
                    -> IO (Either Net.Query.AcquireFailure result)
queryNodeLocalState connctInfo mpoint query = do
    resultVar <- newEmptyTMVarIO
    connectToLocalNode
      connctInfo
      LocalNodeClientProtocols {
        localChainSyncClient    = NoLocalChainSyncClient,
        localStateQueryClient   = Just (singleQuery mpoint resultVar),
        localTxSubmissionClient = Nothing
      }
    atomically (takeTMVar resultVar)
```